src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAuthenticationManager.java

```diff
@@ -496,13 +503,31 @@ public RegisteredUser createUserSession(final HttpServletRequest request, final
      * @param rememberMe - Boolean to indicate whether or not this cookie expiry duration should be long or short
      * @return the request and response will be modified and the original userDO will be returned for convenience.
      */
-    public RegisteredUser createIncompleteLoginUserSession(final HttpServletRequest request, final HttpServletResponse response,
-                                            final RegisteredUser user, final boolean rememberMe) {
-        this.createSession(request, response, user, true, rememberMe);
+    public RegisteredUser createUserSessionWithCaveats(final HttpServletRequest request, final HttpServletResponse response,
+                                                       final RegisteredUser user, final Set<AuthenticationCaveat> authenticationCaveats, final boolean rememberMe) {
+        this.createSession(request, response, user, authenticationCaveats, rememberMe);
         return user;
     }
 
-    /**
+    public RegisteredUser removeCaveatFromUserSession(final HttpServletRequest request, final HttpServletResponse response,
+                                            final RegisteredUser user,
+                                            final AuthenticationCaveat caveatToRemove) throws InvalidSessionException,
+            IOException {
+        Map<String, String> session = getSegueSessionFromRequest(request);
+        String caveats = session.get(SESSION_CAVEATS);
+
+        if (null == caveats) {
+            return null;
+        }
+
+        ArrayList<String> caveatFlags = serializationMapper.readValue(caveats, new TypeReference<ArrayList<String>>(){});
+        caveatFlags.remove(caveatToRemove.toString());
```

Comment:
Should we do something if the session did not have such a caveat? If we were in that situation, something would be very wrong and it would seem worthwhile noting it?

`.remove()` returns false if nothing was removed, so that might be an option.

Reply:
Agreed, this shouldn't happen so I'll log it as a warning.